File: webdriver/branches/chrome/chrome/src/cpp/chromeelement.cc
#include "chromedriver.h"
#include "chromeelement.h"
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <wchar.h>

ChromeElement::~ChromeElement() {
}

void ChromeElement::propagate(ChromeElement* element) {
	element->driver_ = this->driver_;
	element->base_script_ = this->base_script_;
	element->org_query_ = this->org_query_;
	element->query_type_ = this->query_type_;
	element->element_id_ = this->element_id_;
	element->element_type_ = this->element_type_;
}

int ChromeElement::submit() {
  std::wstring jscript;
  driver_->createSubmitScript(jscript, base_script_);
  return driver_->domGetVoid(jscript);
}

int ChromeElement::clear() {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L".clear()");
  return driver_->domGetVoid(jscript);
}

int ChromeElement::click() {
  std::wstring jscript;
  driver_->createClickScript(jscript, base_script_);
  return driver_->domGetVoid(jscript);
}

int ChromeElement::setSelected() {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L".selected=true");
  return driver_->domGetVoid(jscript);
}

int ChromeElement::isEnabled(int* selected) {
  return SUCCESS;
}

int ChromeElement::isSelected(int* selected) {
  return SUCCESS;
}

int ChromeElement::isDisplayed(int* displayed) {
  return SUCCESS;
}

int ChromeElement::toggle(int* toReturn) {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L".checked=true");
  return driver_->domGetVoid(jscript);
}

int ChromeElement::sendKeys(const std::wstring keys) {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L".focus()");
  driver_->domGetVoid(jscript);

  driver_->dg(jscript.c_str());
  driver_->dg(keys.c_str());

  ::sendKeys((void *)&window_handle_, keys.c_str(), 0);
  return SUCCESS;
}

std::wstring ChromeElement::getAttribute(const std::wstring name) {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L"['");
  jscript.append(name.c_str());
  jscript.append(L"']");
  return driver_->domGetString(jscript);
}

std::wstring ChromeElement::getValueOfCssProperty(const std::wstring name) {
  std::wstring jscript = base_script_.c_str();
  if (name == L"className") {
    jscript.append(L".className");
  } else {
    jscript.append(L".style.");
    jscript.append(name);
  }
  return driver_->domGetString(jscript);
}

std::wstring ChromeElement::getText() {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L".innerText");
  return driver_->domGetString(jscript);
}

std::wstring ChromeElement::getElementName() {
  std::wstring jscript = base_script_.c_str();
  jscript.append(L".nodeName");
  return driver_->domGetString(jscript);
}

int ChromeElement::getLocation(long* left, long* top) {
  return SUCCESS;
}

int ChromeElement::getSize(long* width, long* height) {
  return SUCCESS;
}

File: webdriver/branches/chrome/chrome/src/java/org/openqa/selenium/chromium/ChromeElement.java
package org.openqa.selenium.chromium;

import static org.openqa.selenium.chromium.ExportedWebDriver.SUCCESS;

import org.openqa.selenium.chromium.ExportedWebDriver.StringWrapper;
import org.openqa.selenium.internal.Locatable;
import org.openqa.selenium.By;
import org.openqa.selenium.RenderedWebElement;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;

import com.sun.jna.Pointer;
import com.sun.jna.WString;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.NativeLongByReference;
import com.sun.jna.ptr.PointerByReference;

import java.awt.Dimension;
import java.awt.Point;
import java.util.ArrayList;
import java.util.List;

public class ChromeElement implements RenderedWebElement, SearchContext, Locatable {
  private ExportedWebDriver lib;
  private Pointer driver;
  private Pointer element;

  public static List<WebElement> createCollection(
      ExportedWebDriver lib, Pointer driver, Pointer collection) {
    return new ElementCollection(lib, driver, collection).toList();
  }

  public ChromeElement(ExportedWebDriver lib, Pointer driver, Pointer element) {
    this.lib = lib;
    this.driver = driver;
    this.element = element;
  }

  public Dimension getSize() {
    NativeLongByReference width = new NativeLongByReference();
    NativeLongByReference height = new NativeLongByReference();
    int result = lib.wdGetSize(element, width, height);
    if (result != SUCCESS) {
      throw new RuntimeException("Failed to find the size of the element.");
    }
    return new Dimension(width.getValue().intValue(), height.getValue().intValue());
  }

  public Point getLocation() {
    NativeLongByReference left = new NativeLongByReference();
    NativeLongByReference top = new NativeLongByReference();
    int result = lib.wdGetLocation(element, left, top);
    if (result != SUCCESS) {
      throw new RuntimeException("Failed to find location.");
    }
    return new Point(left.getValue().intValue(), top.getValue().intValue());
  }

  // --------------------------------------------------------------------------
  // Implements WebElement
  // --------------------------------------------------------------------------
  public void click() {
    lib.wdClick(element);
    try { Thread.sleep(100); } catch(Exception e) {}
  }

  public void submit() {
    lib.wdSubmit(element);
    try { Thread.sleep(100); } catch(Exception e) {}
  }

  public String getValue() {
    return getAttribute("value");
  }

  public String getElementName() {
    PointerByReference retValue = new PointerByReference();
    int result = lib.wdGetElementName(element, retValue);
    if (result != SUCCESS) {
      throw new RuntimeException("Failed to get element name");
    }
    return new StringWrapper(lib, retValue).toString();
  }

  public String getAttribute(String name) {
    PointerByReference retValue = new PointerByReference();
    int result = lib.wdGetAttribute(element, new WString(name), retValue);
    if (result != SUCCESS) {
      throw new RuntimeException("Failed to attribute: " + name);
    }
    return new StringWrapper(lib, retValue).toString();
  }

  public String getValueOfCssProperty(String propertyName) {
    PointerByReference retValue = new PointerByReference();
    int result = lib.wdGetValueOfCssProperty(element, new WString(propertyName), retValue);
    if (result != SUCCESS) {
      throw new RuntimeException("Failed to css property: " + propertyName);
    }
    return new StringWrapper(lib, retValue).toString();
  }

  public String getText() {
    PointerByReference retValue = new PointerByReference();
    int result = lib.wdGetText(element, retValue);
    if (result != SUCCESS) {
      throw new RuntimeException("Failed to get text");
    }
    return new StringWrapper(lib, retValue).toString();
  }

  public void dragAndDropOn(RenderedWebElement element) {
    // TODO(amitabh): Implement this.
  }

  public void dragAndDropBy(int moveRightBy, int moveDownBy) {
    // TODO(amitabh): Implement this.
  }

  public boolean isDisplayed() {
    IntByReference retValue = new IntByReference();
    lib.wdIsDisplayed(element, retValue);
    return (retValue.getValue() == 1);
  }

  // --------------------------------------------------------------------------
  // Implements SearchContext.
  // --------------------------------------------------------------------------
  public List<WebElement> findElements(By by) {
    return new Finder(lib, driver).findElements(by);
  }

  public WebElement findElement(By by) {
    return new Finder(lib, driver).findElement(by);
  }

  public void sendKeys(CharSequence... keysToSend) {
    StringBuilder builder = new StringBuilder();
    for (CharSequence seq : keysToSend) {
      builder.append(seq);
    }
    lib.wdSendKeys(element, new WString(builder.toString()));
  }

  public void clear() {
    lib.wdClear(element);
  }

  public boolean toggle() {
    IntByReference retValue = new IntByReference();
    lib.wdToggle(element, retValue);
    return (retValue.getValue() == 1);
  }

  public boolean isSelected() {
    IntByReference retValue = new IntByReference();
    lib.wdIsSelected(element, retValue);
    return (retValue.getValue() == 1);
  }

  public void setSelected() {
    lib.wdSetSelected(element);
  }

  public boolean isEnabled() {
    IntByReference retValue = new IntByReference();
    lib.wdIsEnabled(element, retValue);
    return (retValue.getValue() == 1);
  }

  // --------------------------------------------------------------------------
  // Implements Locatable
  // --------------------------------------------------------------------------
  public Point getLocationOnScreenOnceScrolledIntoView() {
    // TODO(amitabh): Implement this.
    return new Point(0, 0);
  }

  // --------------------------------------------------------------------------
  // ElementCollection implementation.
  // --------------------------------------------------------------------------
  private static class ElementCollection {
    private final List<WebElement> elements;

    public ElementCollection(ExportedWebDriver lib, Pointer driver, Pointer collection) {
      elements = extractElements(lib, driver, collection);
    }

    public List<WebElement> toList() {
      return elements;
    }

    private List<WebElement> extractElements(ExportedWebDriver lib, Pointer driver, Pointer collection) {
      IntByReference lenReturn = new IntByReference();
      lib.wdGetCollectionLength(collection, lenReturn);

      int length = lenReturn.getValue();
      List<WebElement> toReturn = new ArrayList<WebElement>(length);
      for (int i = 0; i < length; i++) {
        PointerByReference elemRef = new PointerByReference();
        toReturn.add(i, new ChromeElement(lib, driver, elemRef.getValue()));
      }

      // Free the list in the driver.
      lib.wdFreeElementCollection(collection);
      return toReturn;
    }
  }
}
